hubblestack_nova/modules/cve_scan_v2.py

```diff
@@ -0,0 +1,283 @@
+            log.error('The results weren\'t able to be cached')
+                    
+    ret = {'Success':[], 'Failure':[]}   
+    
+    affected_pkgs = _get_cve_vulnerabilities(master_json)
+    local_pkgs = __salt__['pkg.list_pkgs'](versions_as_list=True)
+    
+    for pkg_obj in affected_pkgs:
+##TODO: for efficiency, should switch the loop to go through just the local_pkgs and check if it's in the affected packages.. 
+        if pkg_obj.get_pkg() in local_pkgs:
+            affected_version = pkg_obj.get_version()
+            for local_version in local_pkgs[pkg_obj.get_pkg()]:
+                if _is_vulnerable(local_version, affected_version, pkg_obj.get_operator()):
+                    ret['Failure'].append(pkg_obj.report())
+                else:
+                    ret['Success'].append(pkg_obj.get_pkg())
+        
+        else:
+            ret['Success'].append(pkg_obj.get_pkg())
+    return ret
+def _get_cve_vulnerabilities(query_results):
+    '''
+    Returns list of vulnerable package objects.
+    ### TODO ### return map of pkg->pkg_obj for more efficient loop structure
+    '''
+    
+    vulnerable_pkgs = []
+
+    # Make sure query was successful
+    if query_results['result'].lower() != 'ok':
+        return
+    
+    # Get os version to only add vulnerabilites that apply to local system 
+    osmajorrelease = str(__grains__['osrelease']).split('.')[0]
+
+    for report in query_results['data']['search']:
+        
+        #data:search
+        reporter = report['_source']['reporter']
+        cve_list = report['_source']['cvelist']
+        href = report['_source']['href']
+        score = report['_source']['cvss']['score']
+        
+        for pkg in report['_source']['affectedPackage']:
+            #data:search:_source:affectedPackages
+            
+            if pkg['OSVersion'] in ['any', osmajorrelease]: # Check if os version matches grains
+                pkg_obj = vulnerablePkg(pkg['packageName'], pkg['packageVersion'], score, pkg['operator'], reporter, href, cve_list)
+                vulnerable_pkgs.append(pkg_obj)   
+    return vulnerable_pkgs
+
+def _is_vulnerable(local_version, affected_version, operator):
+    '''
+    Given two version strings, and operator
+        returns whether the package is vulnerable or not.
+    '''
+    # Get rid of prefix if version number has one, ex '1:3.4.52'
+    if ':' in local_version:
+         local_version = local_version[local_version.index(':')+1:]
+
+    # In order to do compare LooseVersions, eliminate trailing 'el<#>'
+    if re.search('.el\d$', affected_version):
+        affected_version = affected_version[:-4]
+    if re.search('.el\d$', affected_version):
+        affected_version = affected_version[:-4]
+
+    #Compare from higher order to lower order based on '-' split.
+    local_version_split = local_version.split('-')
+    affected_version_split = affected_version.split('-')
+
+    for order_index in range(len(local_version_split)):
```

Comment:
I thought that and was considering doing 

```python
for order_index in range(min([len(local_version_split), len(affected_version_split)])):
    ...
```
but I don't know if we want to run the check as normal in this case. If the affected and local version numbers aren't matching formats, i.e. they don't have the same number of '-' , then we probably want to just report that the version numbers don't match and someone should manually take a look if the package is vulnerable or not. LooseVersion isn't the greatest and we don't want to report that a pkg is safe because LooseVersion reports it is incorrectly. Does that make sense? I'm not sure how hard it would be to make our own version comparison module to fix these cases, but maybe we talk about it.